v10.0.0-beta2: After Backup+Restore: Read and write fails without throwing exceptions

Hi!

First, thanks for your great library!

I am currently testing 10.0.0-beta2 on Android.

Steps to reproduce:

1. Store some item using FlutterSecureStorage
2. Do a backup and restore of the app (on the same device), e.g. using [these instructions](https://developer.android.com/identity/data/testingbackup#Preparing)
3. Open the app again; the library fails to read the stored item (it returns null instead; would be okay, since the secret/key is probably missing) and fails to write any other item without throwing any exceptions (definitely not okay).

I know, that the FlutterSecureStorage files should be excluded from the backup, but I argue that at least an exception should be thrown to the user of the library if the store operation does not succeed; alternatively, the library should fix the broken state (missing/wrong key) itself.

In fact, there is an exception internally that is swallowed [here](https://github.com/juliansteenbakker/flutter_secure_storage/blob/141573a3b91386a696ea2db891282c7b099225fd/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java#L111)

The excerpt of logcat:

```
01-08 17:35:18.110     0     0 I trusty  : WARN: system/keymint/ta/src/lib.rs:1190 failing OperationFinish request with error Hal(VerificationFailed, "system/keymint/boringssl/src/aes.rs:276: failed to finalize: ErrorStack([])")
01-08 17:35:18.109   532   545 W keymint-hal-trusty: kmr_hal: HAL: command OperationFinish failed: -30
01-08 17:35:18.110   529  9853 E keystore2: system/security/keystore2/src/error.rs:174 - system/security/keystore2/src/operation.rs:848: KeystoreOperation::finish
01-08 17:35:18.110   529  9853 E keystore2: 
01-08 17:35:18.110   529  9853 E keystore2: Caused by:
01-08 17:35:18.110   529  9853 E keystore2:     0: system/security/keystore2/src/operation.rs:429: Finish failed.
01-08 17:35:18.110   529  9853 E keystore2:     1: Error::Km(r#VERIFICATION_FAILED)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid: EncryptedSharedPreferences initialization failed
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid: javax.crypto.AEADBadTagException
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:632)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at javax.crypto.Cipher.doFinal(Cipher.java:2132)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.google.crypto.tink.integration.android.AndroidKeystore$AeadImpl.decrypt(AndroidKeystore.java:177)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.decrypt(AndroidKeystoreAesGcm.java:63)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.google.crypto.tink.KeysetHandle.decrypt(KeysetHandle.java:954)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.google.crypto.tink.KeysetHandle.readWithAssociatedData(KeysetHandle.java:835)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.google.crypto.tink.LegacyKeysetSerialization.parseEncryptedKeyset(LegacyKeysetSerialization.java:63)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.readMasterkeyDecryptAndParseKeyset(AndroidKeysetManager.java:387)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.build(AndroidKeysetManager.java:308)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.it_nomads.fluttersecurestorage.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:158)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.it_nomads.fluttersecurestorage.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:124)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.it_nomads.fluttersecurestorage.FlutterSecureStorage.initializeEncryptedSharedPreferencesManager(FlutterSecureStorage.java:130)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.it_nomads.fluttersecurestorage.FlutterSecureStorage.getEncryptedSharedPreferences(FlutterSecureStorage.java:108)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.it_nomads.fluttersecurestorage.FlutterSecureStorage.write(FlutterSecureStorage.java:54)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin$MethodRunner.handleWrite(FlutterSecureStoragePlugin.java:118)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin$MethodRunner.handleMethodCall(FlutterSecureStoragePlugin.java:92)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin$MethodRunner.run(FlutterSecureStoragePlugin.java:80)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.os.Handler.handleCallback(Handler.java:991)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.os.Handler.dispatchMessage(Handler.java:102)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.os.Looper.loopOnce(Looper.java:232)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.os.Looper.loop(Looper.java:317)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.os.HandlerThread.run(HandlerThread.java:85)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid: Caused by: android.security.KeyStoreException: Signature/MAC verification failed (internal Keystore code: -30 message: system/security/keystore2/src/operation.rs:848: KeystoreOperation::finish
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid: 
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid: Caused by:
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:     0: system/security/keystore2/src/operation.rs:429: Finish failed.
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:     1: Error::Km(r#VERIFICATION_FAILED)) (public error code: 10 internal Keystore code: -30)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:416)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:79)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.security.KeyStoreOperation.finish(KeyStoreOperation.java:132)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.finish(KeyStoreCryptoOperationChunkedStreamer.java:228)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:181)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.security.keystore2.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal(AndroidKeyStoreAuthenticatedAESCipherSpi.java:396)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:624)
01-08 17:35:18.114 14738 14787 E SecureStorageAndroid:  ... 21 more
```

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java
@@ -23,24 +23,45 @@ import java.util.Map;
 
 public class FlutterSecureStorage {
 
-    private static final String TAG = "SecureStorageAndroid";
+    private static final String TAG = "FlutterSecureStorage";
     private static final Charset CHARSET = StandardCharsets.UTF_8;
     private static final String DEFAULT_PREF_NAME = "FlutterSecureStorage";
     private static final String DEFAULT_KEY_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBhIHNlY3VyZSBzdG9yYWdlCg";
-    @NonNull
-    private final Context applicationContext;
+    private static final String PREF_OPTION_NAME = "sharedPreferencesName";
+    private static final String PREF_OPTION_PREFIX = "preferencesKeyPrefix";
+    private static final String PREF_OPTION_DELETE_ON_FAILURE = "resetOnError";
+    private static final String PREF_KEY_MIGRATED = "preferencesMigrated";
     @NonNull
     private final SharedPreferences encryptedPreferences;
-    private final Map<String, Object> options;
-    private String sharedPreferencesName = DEFAULT_PREF_NAME;
+    @NonNull
     private String preferencesKeyPrefix = DEFAULT_KEY_PREFIX;
 
-
     public FlutterSecureStorage(Context context, Map<String, Object> options) throws GeneralSecurityException, IOException {
-        this.applicationContext = context.getApplicationContext();
-        this.options = options;
-        ensureOptions();
-        encryptedPreferences = getEncryptedSharedPreferences();
+        String sharedPreferencesName = DEFAULT_PREF_NAME;
+        if (options.containsKey(PREF_OPTION_NAME)) {
+            var value = options.get(PREF_OPTION_NAME);
+            if (value instanceof String && !((String) value).isEmpty()) {
+                sharedPreferencesName = (String) value;
+            }
+        }
+
+        if (options.containsKey(PREF_OPTION_PREFIX)) {
+            var value = options.get(PREF_OPTION_PREFIX);
+            if (value instanceof String && !((String) value).isEmpty()) {
+                preferencesKeyPrefix = (String) value;
+            }
+        }
+
+        boolean deleteOnFailure = false;
+
+        if (options.containsKey(PREF_OPTION_DELETE_ON_FAILURE)) {
+            var value = options.get(PREF_OPTION_DELETE_ON_FAILURE);
+            if (value instanceof String) {
+                deleteOnFailure = Boolean.parseBoolean((String) value);
+            }
+        }
+
+        encryptedPreferences = getEncryptedSharedPreferences(deleteOnFailure, options, context.getApplicationContext(), sharedPreferencesName);
     }
 
     public boolean containsKey(String key) {
@@ -81,34 +102,34 @@ public class FlutterSecureStorage {
         return preferencesKeyPrefix + "_" + key;
     }
 
-    private void ensureOptions() {
-        sharedPreferencesName = options.containsKey("sharedPreferencesName") && options.get("sharedPreferencesName") instanceof String
-                ? (String) options.get("sharedPreferencesName")
-                : DEFAULT_PREF_NAME;
-
-        preferencesKeyPrefix = options.containsKey("preferencesKeyPrefix") && options.get("preferencesKeyPrefix") instanceof String
-                ? (String) options.get("preferencesKeyPrefix")
-                : DEFAULT_KEY_PREFIX;
-    }
-
-    private SharedPreferences getEncryptedSharedPreferences() throws GeneralSecurityException, IOException {
+    private SharedPreferences getEncryptedSharedPreferences(boolean deleteOnFailure, Map<String, Object> options, Context context, String sharedPreferencesName) throws GeneralSecurityException, IOException {
         try {
-            final SharedPreferences encryptedPreferences = initializeEncryptedSharedPreferencesManager(applicationContext);
-            migrateToEncryptedPreferences(encryptedPreferences);
+            final SharedPreferences encryptedPreferences = initializeEncryptedSharedPreferencesManager(context, sharedPreferencesName);
+            boolean migrated = encryptedPreferences.getBoolean(PREF_KEY_MIGRATED, false);
+            if (!migrated) {
+                migrateToEncryptedPreferences(context, sharedPreferencesName, encryptedPreferences, deleteOnFailure, options);
+            }
             return encryptedPreferences;
-        } catch (Exception e) {
-            Log.w(TAG, "EncryptedSharedPreferences initialization failed, resetting storage", e);
-            applicationContext.getSharedPreferences(sharedPreferencesName, Context.MODE_PRIVATE).edit().clear().apply();
+        } catch (GeneralSecurityException | IOException e) {
+
+            if (!deleteOnFailure) {
+                Log.w(TAG, "initialization failed, resetOnError false, so throwing exception.", e);
+                throw e;
+            }
+            Log.w(TAG, "initialization failed, resetting storage", e);
+
+            context.getSharedPreferences(sharedPreferencesName, Context.MODE_PRIVATE).edit().clear().apply();
+
             try {
-                return initializeEncryptedSharedPreferencesManager(applicationContext);
+                return initializeEncryptedSharedPreferencesManager(context, sharedPreferencesName);
             } catch (Exception f) {
-                Log.e(TAG, "EncryptedSharedPreferences initialization after reset failed", e);
+                Log.e(TAG, "initialization after reset failed", f);
                 throw f;
             }
         }
     }
 
-    private SharedPreferences initializeEncryptedSharedPreferencesManager(Context context) throws GeneralSecurityException, IOException {
+    private SharedPreferences initializeEncryptedSharedPreferencesManager(Context context, String sharedPreferencesName) throws GeneralSecurityException, IOException {
         MasterKey masterKey = new MasterKey.Builder(context)
                 .setKeyGenParameterSpec(new KeyGenParameterSpec.Builder(
                         MasterKey.DEFAULT_MASTER_KEY_ALIAS,
@@ -128,17 +149,18 @@ public class FlutterSecureStorage {
         );
     }
 
-    private void migrateToEncryptedPreferences(SharedPreferences target) {
-        SharedPreferences source = applicationContext.getSharedPreferences(sharedPreferencesName, Context.MODE_PRIVATE);
+    private void migrateToEncryptedPreferences(Context context, String sharedPreferencesName, SharedPreferences target, boolean deleteOnFailure, Map<String, Object> options) {
+        SharedPreferences source = context.getSharedPreferences(sharedPreferencesName, Context.MODE_PRIVATE);
+
+        Map<String, ?> sourceEntries = source.getAll();
+        if (sourceEntries.isEmpty()) return;
+
+        int succesfull = 0;
+        int failed = 0;
 
         try {
-            StorageCipher cipher = new StorageCipherFactory(source, options).getSavedStorageCipher(applicationContext);
+            StorageCipher cipher = new StorageCipherFactory(source, options).getSavedStorageCipher(context);
 
-            Map<String, ?> sourceEntries = source.getAll();
-            if (sourceEntries.isEmpty()) return;
-
-            int succesfull = 0;
-            int failed = 0;
             for (Map.Entry<String, ?> entry : sourceEntries.entrySet()) {
                 String key = entry.getKey();
                 Object value = entry.getValue();
@@ -151,6 +173,10 @@ public class FlutterSecureStorage {
                     } catch (Exception e) {
                         Log.e(TAG, "Migration failed for key: " + key, e);
                         failed++;
+
+                        if (deleteOnFailure) {
+                            source.edit().remove(key).apply();
+                        }
                     }
                 }
             }
@@ -158,11 +184,23 @@ public class FlutterSecureStorage {
             if (succesfull > 0) {
                 Log.i(TAG, "Successfully migrated " + succesfull + " keys.");
             }
+
             if (failed > 0) {
-                Log.i(TAG, "Failed to migrate " + failed + " keys.");
+                Log.w(TAG, "Failed to migrate " + failed + " keys.");
             }
+
+            if (failed == 0 || deleteOnFailure) {
+                target.edit().putBoolean(PREF_KEY_MIGRATED, true).apply();
+            }
+
         } catch(Exception e) {
             Log.e(TAG, "Migration failed due to initialisation error.", e);
+
+            // If a failure has occurred during StorageCipher initialization, set migrated to true
+            // so migration is not run again
+            if (deleteOnFailure) {
+                target.edit().putBoolean(PREF_KEY_MIGRATED, true).apply();
+            }
         }
     }
 

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java
@@ -31,51 +31,45 @@ public class FlutterSecureStorage {
 
     private String sharedPreferencesName = DEFAULT_PREF_NAME;
     private String preferencesKeyPrefix = DEFAULT_KEY_PREFIX;
-    private SharedPreferences encryptedPreferences;
+    private final SharedPreferences encryptedPreferences;
 
-    public FlutterSecureStorage(Context context, Map<String, Object> options) {
+    public FlutterSecureStorage(Context context, Map<String, Object> options) throws GeneralSecurityException, IOException {
         this.applicationContext = context.getApplicationContext();
         this.options = options;
         ensureOptions();
-        getEncryptedSharedPreferences();
+        encryptedPreferences = getEncryptedSharedPreferences();
     }
 
     public boolean containsKey(String key) {
-        SharedPreferences preferences = getEncryptedSharedPreferences();
-        return preferences != null && preferences.contains(addPrefixToKey(key));
+        return encryptedPreferences != null && encryptedPreferences.contains(addPrefixToKey(key));
     }
 
     public String read(String key) {
-        SharedPreferences preferences = getEncryptedSharedPreferences();
-        return preferences != null ? preferences.getString(addPrefixToKey(key), null) : null;
+        return encryptedPreferences != null ? encryptedPreferences.getString(addPrefixToKey(key), null) : null;
     }
 
     public void write(String key, String value) {
-        SharedPreferences preferences = getEncryptedSharedPreferences();
-        if (preferences != null) {
-            preferences.edit().putString(addPrefixToKey(key), value).apply();
+        if (encryptedPreferences != null) {
+            encryptedPreferences.edit().putString(addPrefixToKey(key), value).apply();
         }
     }
 
     public void delete(String key) {
-        SharedPreferences preferences = getEncryptedSharedPreferences();
-        if (preferences != null) {
-            preferences.edit().remove(addPrefixToKey(key)).apply();
+        if (encryptedPreferences != null) {
+            encryptedPreferences.edit().remove(addPrefixToKey(key)).apply();
         }
     }
 
     public void deleteAll() {
-        SharedPreferences preferences = getEncryptedSharedPreferences();
-        if (preferences != null) {
-            preferences.edit().clear().apply();
+        if (encryptedPreferences != null) {
+            encryptedPreferences.edit().clear().apply();
         }
     }
 
     public Map<String, String> readAll() {
-        SharedPreferences preferences = getEncryptedSharedPreferences();
         Map<String, String> result = new HashMap<>();
-        if (preferences != null) {
-            Map<String, ?> allEntries = preferences.getAll();
+        if (encryptedPreferences != null) {
+            Map<String, ?> allEntries = encryptedPreferences.getAll();
             for (Map.Entry<String, ?> entry : allEntries.entrySet()) {
                 String key = entry.getKey();
                 Object value = entry.getValue();
@@ -102,18 +96,21 @@ public class FlutterSecureStorage {
                 : DEFAULT_KEY_PREFIX;
     }
 
-    private SharedPreferences getEncryptedSharedPreferences() {
-        if (encryptedPreferences == null) {
-            try {
-                encryptedPreferences = initializeEncryptedSharedPreferencesManager(applicationContext);
-                migrateToEncryptedPreferences(encryptedPreferences);
-            } catch (Exception e) {
-                Log.e(TAG, "EncryptedSharedPreferences initialization failed", e);
-            }
-        } else {
+    private SharedPreferences getEncryptedSharedPreferences() throws GeneralSecurityException, IOException {
+        try {
+            final SharedPreferences encryptedPreferences = initializeEncryptedSharedPreferencesManager(applicationContext);
+            migrateToEncryptedPreferences(encryptedPreferences);
             return encryptedPreferences;
+        } catch (Exception e) {
+            Log.w(TAG, "EncryptedSharedPreferences initialization failed, resetting storage", e);
+            applicationContext.getSharedPreferences(sharedPreferencesName, Context.MODE_PRIVATE).edit().clear().apply();
+            try {
+                return initializeEncryptedSharedPreferencesManager(applicationContext);
+            } catch (Exception f) {
+                Log.e(TAG, "EncryptedSharedPreferences initialization after reset failed", e);
+                throw f;
+            }
         }
-        return null;
     }
 
     private SharedPreferences initializeEncryptedSharedPreferencesManager(Context context) throws GeneralSecurityException, IOException {

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java
@@ -7,6 +7,8 @@ import android.security.keystore.KeyProperties;
 import android.util.Base64;
 import android.util.Log;
 
+import androidx.annotation.NonNull;
+
 import com.it_nomads.fluttersecurestorage.ciphers.StorageCipher;
 import com.it_nomads.fluttersecurestorage.ciphers.StorageCipherFactory;
 import com.it_nomads.fluttersecurestorage.crypto.EncryptedSharedPreferences;
@@ -25,13 +27,14 @@ public class FlutterSecureStorage {
     private static final Charset CHARSET = StandardCharsets.UTF_8;
     private static final String DEFAULT_PREF_NAME = "FlutterSecureStorage";
     private static final String DEFAULT_KEY_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBhIHNlY3VyZSBzdG9yYWdlCg";
-
+    @NonNull
     private final Context applicationContext;
+    @NonNull
+    private final SharedPreferences encryptedPreferences;
     private final Map<String, Object> options;
-
     private String sharedPreferencesName = DEFAULT_PREF_NAME;
     private String preferencesKeyPrefix = DEFAULT_KEY_PREFIX;
-    private final SharedPreferences encryptedPreferences;
+
 
     public FlutterSecureStorage(Context context, Map<String, Object> options) throws GeneralSecurityException, IOException {
         this.applicationContext = context.getApplicationContext();
@@ -41,42 +44,34 @@ public class FlutterSecureStorage {
     }
 
     public boolean containsKey(String key) {
-        return encryptedPreferences != null && encryptedPreferences.contains(addPrefixToKey(key));
+        return encryptedPreferences.contains(addPrefixToKey(key));
     }
 
     public String read(String key) {
-        return encryptedPreferences != null ? encryptedPreferences.getString(addPrefixToKey(key), null) : null;
+        return encryptedPreferences.getString(addPrefixToKey(key), null);
     }
 
     public void write(String key, String value) {
-        if (encryptedPreferences != null) {
-            encryptedPreferences.edit().putString(addPrefixToKey(key), value).apply();
-        }
+        encryptedPreferences.edit().putString(addPrefixToKey(key), value).apply();
     }
 
     public void delete(String key) {
-        if (encryptedPreferences != null) {
-            encryptedPreferences.edit().remove(addPrefixToKey(key)).apply();
-        }
+        encryptedPreferences.edit().remove(addPrefixToKey(key)).apply();
     }
 
     public void deleteAll() {
-        if (encryptedPreferences != null) {
-            encryptedPreferences.edit().clear().apply();
-        }
+        encryptedPreferences.edit().clear().apply();
     }
 
     public Map<String, String> readAll() {
         Map<String, String> result = new HashMap<>();
-        if (encryptedPreferences != null) {
-            Map<String, ?> allEntries = encryptedPreferences.getAll();
-            for (Map.Entry<String, ?> entry : allEntries.entrySet()) {
-                String key = entry.getKey();
-                Object value = entry.getValue();
-                if (key.startsWith(preferencesKeyPrefix) && value instanceof String) {
-                    String originalKey = key.replaceFirst(preferencesKeyPrefix + "_", "");
-                    result.put(originalKey, (String) value);
-                }
+        Map<String, ?> allEntries = encryptedPreferences.getAll();
+        for (Map.Entry<String, ?> entry : allEntries.entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
+            if (key.startsWith(preferencesKeyPrefix) && value instanceof String) {
+                String originalKey = key.replaceFirst(preferencesKeyPrefix + "_", "");
+                result.put(originalKey, (String) value);
             }
         }
         return result;

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStorage.java
@@ -156,7 +156,7 @@ public class FlutterSecureStorage {
             }
 
             if (succesfull > 0) {
-                Log.i(TAG, "Succesfully migrated " + succesfull + " keys.");
+                Log.i(TAG, "Successfully migrated " + succesfull + " keys.");
             }
             if (failed > 0) {
                 Log.i(TAG, "Failed to migrate " + failed + " keys.");

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStoragePlugin.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStoragePlugin.java
@@ -12,7 +12,6 @@ import java.util.HashMap;
 import java.util.Map;
 
 import io.flutter.embedding.engine.plugins.FlutterPlugin;
-import io.flutter.plugin.common.BinaryMessenger;
 import io.flutter.plugin.common.MethodCall;
 import io.flutter.plugin.common.MethodChannel;
 import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
@@ -29,7 +28,11 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
     @Override
     public void onAttachedToEngine(FlutterPluginBinding binding) {
         this.binding = binding;
-        initInstance(binding.getBinaryMessenger());
+        workerThread = new HandlerThread("fluttersecurestorage.worker");
+        workerThread.start();
+        workerThreadHandler = new Handler(workerThread.getLooper());
+        channel = new MethodChannel(binding.getBinaryMessenger(), "plugins.it_nomads.com/flutter_secure_storage");
+        channel.setMethodCallHandler(this);
     }
 
     @Override
@@ -45,20 +48,11 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
         secureStorage = null;
     }
 
-    private void initInstance(BinaryMessenger messenger) {
-        workerThread = new HandlerThread("fluttersecurestorage.worker");
-        workerThread.start();
-        workerThreadHandler = new Handler(workerThread.getLooper());
-        channel = new MethodChannel(messenger, "plugins.it_nomads.com/flutter_secure_storage");
-        channel.setMethodCallHandler(this);
-        initSecureStorage(null);
-    }
-
-    private boolean initSecureStorage(Result result) {
+    private boolean initSecureStorage(Result result, Map<String, Object> options) {
         if (secureStorage != null) return true;
 
         try {
-            secureStorage = new FlutterSecureStorage(binding.getApplicationContext(), new HashMap<>());
+            secureStorage = new FlutterSecureStorage(binding.getApplicationContext(), options);
             return true;
         } catch (Exception e) {
             if (result != null) {
@@ -98,26 +92,33 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
 
         private void handleMethodCall(MethodCall call, Result result) {
             String method = call.method;
-            Map<String, Object> args = extractArguments(call);
+            Map<String, Object> arguments = call.arguments();
 
-            boolean isInitialized = initSecureStorage(result);
+            if (arguments == null) {
+                result.error("InvalidArgument", "No arguments passed to method call", null);
+                return;
+            }
+
+            Map<String, Object> options = extractMapFromObject(arguments.get("options"));
+
+            boolean isInitialized = initSecureStorage(result, options);
             if (!isInitialized) return;
 
             switch (method) {
                 case "write":
-                    handleWrite(args, result);
+                    handleWrite(arguments, result);
                     break;
                 case "read":
-                    handleRead(args, result);
+                    handleRead(arguments, result);
                     break;
                 case "readAll":
                     handleReadAll(result);
                     break;
                 case "containsKey":
-                    handleContainsKey(args, result);
+                    handleContainsKey(arguments, result);
                     break;
                 case "delete":
-                    handleDelete(args, result);
+                    handleDelete(arguments, result);
                     break;
                 case "deleteAll":
                     handleDeleteAll(result);
@@ -140,11 +141,7 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
 
         private void handleRead(Map<String, Object> args, Result result) {
             String key = (String) args.get("key");
-            if (secureStorage.containsKey(key)) {
-                result.success(secureStorage.read(key));
-            } else {
-                result.success(null);
-            }
+            result.success(secureStorage.read(key));
         }
 
         private void handleReadAll(Result result) {
@@ -168,8 +165,11 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
         }
 
         @SuppressWarnings("unchecked")
-        private Map<String, Object> extractArguments(MethodCall call) {
-            return (Map<String, Object>) call.arguments;
+        private Map<String, Object> extractMapFromObject(Object object) {
+            if (!(object instanceof Map)) {
+                return new HashMap<>();
+            }
+            return (Map<String, Object>) object;
         }
 
         private void handleException(Exception e) {

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStoragePlugin.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStoragePlugin.java
@@ -1,10 +1,8 @@
 package com.it_nomads.fluttersecurestorage;
 
-import android.content.Context;
 import android.os.Handler;
 import android.os.HandlerThread;
 import android.os.Looper;
-import android.util.Log;
 
 import androidx.annotation.NonNull;
 
@@ -22,15 +20,16 @@ import io.flutter.plugin.common.MethodChannel.Result;
 
 public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlugin {
 
-    private static final String TAG = "FlutterSecureStoragePl";
     private MethodChannel channel;
     private FlutterSecureStorage secureStorage;
     private HandlerThread workerThread;
     private Handler workerThreadHandler;
+    private FlutterPluginBinding binding;
 
     @Override
     public void onAttachedToEngine(FlutterPluginBinding binding) {
-        initInstance(binding.getBinaryMessenger(), binding.getApplicationContext());
+        this.binding = binding;
+        initInstance(binding.getBinaryMessenger());
     }
 
     @Override
@@ -46,16 +45,30 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
         secureStorage = null;
     }
 
-    private void initInstance(BinaryMessenger messenger, Context context) {
+    private void initInstance(BinaryMessenger messenger) {
+        workerThread = new HandlerThread("fluttersecurestorage.worker");
+        workerThread.start();
+        workerThreadHandler = new Handler(workerThread.getLooper());
+        channel = new MethodChannel(messenger, "plugins.it_nomads.com/flutter_secure_storage");
+        channel.setMethodCallHandler(this);
+        initSecureStorage(null);
+    }
+
+    private boolean initSecureStorage(Result result) {
+        if (secureStorage != null) return true;
+
         try {
-            secureStorage = new FlutterSecureStorage(context, new HashMap<>());
-            workerThread = new HandlerThread("fluttersecurestorage.worker");
-            workerThread.start();
-            workerThreadHandler = new Handler(workerThread.getLooper());
-            channel = new MethodChannel(messenger, "plugins.it_nomads.com/flutter_secure_storage");
-            channel.setMethodCallHandler(this);
+            secureStorage = new FlutterSecureStorage(binding.getApplicationContext(), new HashMap<>());
+            return true;
         } catch (Exception e) {
-            Log.e(TAG, "Plugin initialization failed", e);
+            if (result != null) {
+                result.error(
+                        "RESET_FAILED",  // Error code
+                        "Failed to reset and initialize encrypted preferences", // Error message
+                        e.toString()     // Details (stack trace or additional info)
+                );
+            }
+            return false;
         }
     }
 
@@ -86,6 +99,9 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
         private void handleMethodCall(MethodCall call, Result result) {
             String method = call.method;
             Map<String, Object> args = extractArguments(call);
+
+            boolean isInitialized = initSecureStorage(result);
+            if (!isInitialized) return;
 
             switch (method) {
                 case "write":

--- a/flutter_secure_storage/example/android/gradle/wrapper/gradle-wrapper.properties
+++ b/flutter_secure_storage/example/android/gradle/wrapper/gradle-wrapper.properties
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.9-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.10.2-bin.zip

--- a/flutter_secure_storage/example/android/settings.gradle.kts
+++ b/flutter_secure_storage/example/android/settings.gradle.kts
@@ -19,7 +19,7 @@ pluginManagement {
 
 plugins {
     id("dev.flutter.flutter-plugin-loader") version "1.0.0"
-    id("com.android.application") version "8.7.3" apply false
+    id("com.android.application") version "8.8.0" apply false
     id("org.jetbrains.kotlin.android") version "1.9.25" apply false
 }
 

